feat: Timeline UI

johnmarksilly opened the pull request:
### Timeline UI

Created a new UI for viewing events and live camera views

https://user-images.githubusercontent.com/11224731/149416284-6a1cba74-b048-4c78-8012-2a05c92d4405.mov

To be completed before PR is ready
- [x] Investigate using the recordings framework for faster loading (load segments instead)
- [x] Add config for using new UI
- [x] Update docs
- [x] Handle no events for today
- [x] Load first timeline entry (most recent) and scroll to that event
- [x] Play/Pause + disable arrows on end indices
- [x] Placeholder sizing while video is loading to reduce "buggy" looking element + missing image
- [x] Move timeline while video plays (timeline will be disabled while video is playing)
- [x] Add global padding back to pages (except new route)
- [x] Make sure change complies with dark mode

[Visual Update 02/15/22]

https://user-images.githubusercontent.com/11224731/154198712-4bae4c47-65c0-47b4-99be-d34081c5ffc8.mov

johnmarksill committed:
refactor: moved dialog to prompt to make dialog reusable

johnmarksill committed:
feat: rounded corners on camera cards

johnmarksill committed:
feat: add timeline

johnmarksill committed:
chore: delete rounded edge on cameraimage

johnmarksill committed:
chore: added newlines

johnmarksill committed:
chore: delete unused tailwind config change

johnmarksill committed:
chore: proper formatting

johnmarksilly committed:
chore: add comma back

johnmarksilly committed:
chore: removed unused declarations

NickM-27 commented:
I know it's early but just wanted to say the screenshot looks great!

johnmarksilly committed:
chore: add jsx to prettier

johnmarksilly committed:
feat: add class to hide scrollbar as tailwind doesn't have an option

johnmarksilly committed:
feat: moved events to camera and added logic to track video progress

johnmarksilly committed:
chore: removed background gradient from header

johnmarksilly committed:
chore: moved timeline viewer to new component

johnmarksilly committed:
fix: linting

johnmarksilly committed:
fix: dark mode

johnmarksilly committed:
fix: bug with timeline shifting

johnmarksilly committed:
fix: video not replaced if a new event was queued

johnmarksilly committed:
chore: refactored debug camera and cleaned up cameras_v2

johnmarksilly committed:
fix: added padding back

johnmarksilly committed:
fix: add padding back to cameras page

johnmarksilly committed:
feat: add experimental feature + fix old camera padding

johnmarksilly commented on web/src/components/HistoryViewer.jsx:
```diff
@@ -0,0 +1,117 @@
+  const apiHost = useApiHost();
+  const videoRef = useRef();
+
+  // const beginningOfDay = new Date().setHour(0, 0, 0) / 1000;
+  const beginningOfDay = new Date(new Date().getTime() - 24 * 60 * 60 * 1000);
```
I'd like this to be day of, and potentially if you reach the end, load the next days? Or would a dropdown be better?

johnmarksilly committed:
chore: refactor config naming

johnmarksilly committed:
Revert "safe refactoring (#2552)"

This reverts commit 9edf38347c2215ae865afec2366b83dfc654b5de.

NickM-27 commented:
Looks awesome! Does it also allow viewing parts of the saved recordings that don't have an associated event?

johnmarksilly committed:
fix: fix buggy height between image loads

johnmarksilly commented:
> Looks awesome! Does it also allow viewing parts of the saved recordings that don't have an associated event?

I'm actually working on that, but I think it might be useful to put that in another PR since that's going to take a bit of experimentation. Not everyone uses the recordings so I'd like to figure out a way to integrate it seamlessly without interrupting the experience of those without extra recordings.

I have 24/7 recording so in essence it would replace all the empty space with the recorded content

NickM-27 commented:
> > Looks awesome! Does it also allow viewing parts of the saved recordings that don't have an associated event?
> 
> I'm actually working on that, but I think it might be useful to put that in another PR since that's going to take a bit of experimentation. Not everyone uses the recordings so I'd like to figure out a way to integrate it seamlessly without interrupting the experience of those without extra recordings.
> 
> I have 24/7 recording so in essence it would replace all the empty space with the recorded content
> 
> [EDIT]: This is basically a 1v1 replacement for the old camera page (and moving the last 24 hour events into this page)

Ah I see, that makes sense. I also have 24/7 recording which is why I asked.

johnmarksilly committed:
chore: refactor to reduce re-rendering

johnmarksilly committed:
fix: replace time search with 24 hour

hcgonzalezpr commented:
Looks awesome, almost what I was looking in https://github.com/blakeblackshear/frigate/issues/2230

johnmarksilly committed:
refactor: used hls system instead

johnmarksilly committed:
refactor: rewrote logic for building timeline to support stacking and changed to typescript

johnmarksilly committed:
Revert "Revert "safe refactoring (#2552)""

This reverts commit 6dfcc33eb8ecdd6af1726304d206dc8ae6dacb41.

johnmarksilly commented:
Sorry for such a long delay! I have been extremely sick and didn't have much time to do anything!

Updates:
- refactored the rendering logic in order to support offsetting the timeline blocks vertically for overlapping events
- refactored to typescript because I couldn't handle not knowing if my objects were consistent. 
- misc fixes

There is an outstanding bug with how the overlapping offset logic works that I have to address

johnmarksilly committed:
feat: added color mapping by label

johnmarksilly committed:
refactor: pulled components apart and improved logic

johnmarksilly committed:
refactor: revert quote type in app.jsx

johnmarksilly committed:
chore: revert quotes in birdseye

johnmarksilly committed:
chore: revert quotes in Camera.jsx

johnmarksilly committed:
chore: revert quotes in camermap

johnmarksilly committed:
chore: revert spaces

johnmarksilly committed:
chore: revert fix formatting in camera.jsx

johnmarksilly committed:
chore: formatting

johnmarksilly committed:
chore: fix formatting

johnmarksilly committed:
chore: added docs

johnmarksilly committed:
chore: refactor

johnmarksilly committed:
chore: removed log statement

johnmarksilly committed:
feat: added className to TimelineControls

johnmarksilly committed:
refactor: TimelineBlockView

johnmarksilly committed:
refactor: TimelineBlocks

johnmarksilly committed:
chore: reduce re-renders

johnmarksilly committed:
refactor: add newline to tsconfig.json

johnmarksilly committed:
fix: fix playback button

johnmarksilly commented:
The time has come 🗡️ 

There are some quirks with chrome/android at the moment that I'm still trying to work out. Lots of planned improvements:

- [ ] Loading screens for fetching events
- [ ] Pre-load snapshots that are before and after the marker position for seamless scrubbing
- [ ] Pre-cache timeline events on the backend (similar to Nest) by saving a video for a full day that with 1 frame per minute or something like that
- [ ] Load recording if no events occur at a specified marker time
- [ ] Start video at marker position
- [ ] Don't switch events to an overlapping event while playing an event (this causes the video to load the next event).
- [ ] Refactoring to make things more decoupled. I don't like having to bubble up the play state through the timeline component

Lots more to come. If anyone wants a demo or questions/comments/concerns, post here and maybe I can schedule a Zoom/Discord meeting (I'm in CST)

johnmarksilly committed:
lint: self-closing tags in LiveChip

johnmarksilly committed:
lint: fix import

johnmarksilly committed:
lint: fixes from lint and updates to typescript linting

johnmarksilly committed:
fix: reverted package.json hostname changes

johnmarksilly committed:
fix: tailwindcss to support tsx

johnmarksilly committed:
fix: missing ts files and adjusted tailwind settings to look at ts tailwind util files

johnmarksilly committed:
fix: lint with quotes

johnmarksilly commented:
@blakeblackshear I think this can be re-run. I'm currently deploying to my k8s cluster and the prod build works locally so should be 💯

johnmarksilly commented:
Eff I forgot to run the tests. I'll fix quick.

johnmarksilly committed:
fix: added a test for the prompt and refactored the dialog test

johnmarksilly committed:
fix: dialog import and video height fix

blakeblackshear commented:
Please open a new PR to `release-0.11.0`.

johnmarksilly commented:
Sounds good. Do you not prefer to change the base branch in open PRs? It's a personal preference, but it maintains the history of the PR/comments if you don't close the branch. I've never had a public repo so I don't know if that's normal to do when having collaborators create PRs against forks.
